src/vm/costs/mod.rs

```diff
@@ -134,6 +211,73 @@ impl LimitedCostTracker {
     }
 }
 
+fn parse_cost(cost_function: ClarityCostFunction, eval_result: InterpreterResult<Option<Value>>) -> Result<ExecutionCost> {
+    match eval_result {
+        Ok(Some(Value::Tuple(data))) => {
+            let results = (
+                data.data_map.get("write_length"),
+                data.data_map.get("write_count"),
+                data.data_map.get("runtime"),
+                data.data_map.get("read_length"),
+                data.data_map.get("read_count"),
+            );
+
+            match results {
+                (Some(UInt(write_length)),
+                    Some(UInt(write_count)),
+                    Some(UInt(runtime)),
+                    Some(UInt(read_length)),
+                    Some(UInt(read_count))) => {
+                    Ok(ExecutionCost {
+                        write_length: (*write_length as u64),
+                        write_count: (*write_count as u64),
+                        runtime: (*runtime as u64),
+                        read_length: (*read_length as u64),
+                        read_count: (*read_count as u64),
+                    })
+                },
+                _ => Err(CostErrors::CostComputationFailed("Execution Cost tuple does not contain only UInts".to_string()))
+            }
+        },
+        Ok(Some(_)) => Err(CostErrors::CostComputationFailed("Clarity cost function returned something other than a Cost tuple".to_string())),
+        Ok(None) => Err(CostErrors::CostComputationFailed("Clarity cost function returned nothing".to_string())),
+        Err(e) => Err(CostErrors::CostComputationFailed(format!("Error evaluating result of cost function {}: {}", cost_function.get_name(), e))),
+    }
+}
+
+fn compute_cost(
+    cost_tracker: &mut LimitedCostTracker,
+    cost_function: ClarityCostFunction,
+    input_size: u64) -> Result<ExecutionCost> {
+
+    let mut null_store = NullBackingStore::new();
```

Comment:
The code doesn't seem to be parsing through the cost function code to verify that no cost function does any I/O.  This is necessary because the use of `NullBackingStore` will cause the node to panic if it uses a cost function that attempts any Clarify function that does I/O under the hood.

Can you update the code that parses the cost functions into an AST to return an error if the AST contains any nodes that do I/O (or do fold, map, or filter)?

Reply:
This is captured in #2041